File: DSA/Ansh-Gulati/Day-19/Day-19_Question-2.cpp
#include <iostream>
#include <vector>

using namespace std;

void dfsUtil(int node, const vector<vector<int>>& adj, vector<bool>& visited, vector<int>& result) {
    visited[node] = true;
    result.push_back(node);

    for (int neighbor : adj[node]) {
        if (!visited[neighbor]) {
            dfsUtil(neighbor, adj, visited, result);
        }
    }
}

vector<int> dfs(int start, int nodes, const vector<vector<int>>& adj) {
    vector<bool> visited(nodes, false);
    vector<int> result;
    dfsUtil(start, adj, visited, result);
    return result;
}

int main() {
    int nodes, edges;
    cin >> nodes >> edges;

    vector<vector<int>> adj(nodes);

    for (int i = 0; i < edges; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u - 1].push_back(v - 1); // Assuming nodes are 1-indexed
        adj[v - 1].push_back(u - 1); // Since the graph is undirected
    }

    int start;
    cin >> start;

    vector<int> result = dfs(start - 1, nodes, adj); // Assuming nodes are 1-indexed

    for (int node : result) {
        cout << "Node " << node + 1 << " ";
    }
    cout << endl;

    return 0;
}

File: DSA/Manav-Rajput/Day19_question2.java
package new1;
import java.util.*;

public class DepthFirstSearch {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Read number of nodes and edges
        int nodes = scanner.nextInt();
        int edges = scanner.nextInt();

        // Initialize the adjacency list for the graph (1-based indexing)
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= nodes; i++) {
            graph.add(new ArrayList<>());
        }

        // Read edges
        for (int i = 0; i < edges; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        // Read starting node
        int startNode = scanner.nextInt();
        scanner.close();

        // Perform DFS and get the nodes visited
        List<Integer> visitedNodes = new ArrayList<>();
        boolean[] visited = new boolean[nodes + 1]; // 1-based indexing

        // Start DFS from the specified starting node
        dfs(graph, startNode, visited, visitedNodes);

        // Print nodes visited in DFS order
        System.out.println("Nodes visited in DFS order:");
        for (int node : visitedNodes) {
            System.out.print(node + " ");
        }
        System.out.println();
    }

    private static void dfs(List<List<Integer>> graph, int node, boolean[] visited, List<Integer> visitedNodes) {
        // Mark the current node as visited
        visited[node] = true;
        visitedNodes.add(node);

        // Visit all neighbors of the current node
        for (int neighbor : graph.get(node)) {
            if (!visited[neighbor]) {
                dfs(graph, neighbor, visited, visitedNodes);
            }
        }
    }
}
